index.rst

```diff
@@ -1,57 +1,451 @@
+As a result, these classes are difficult to adapt to new use cases.
+In addition, the lack of a common code base makes these classes difficult to maintain, and the limited type safety makes these classes easy to use incorrectly from both C++ and Python.
+:class:`~lsst.afw.typehandling.GenericMap` attempted to not only serve as a suitable back-end for :class:`~lsst.afw.image.ExposureInfo`, but also address the design problems that prevented the use of :class:`~lsst.daf.base.PropertySet` in the first place.
+
+Currently, :class:`lsst.afw.typehandling.GenericMap` is a component of ``afw`` because it depends, indirectly, on the ``lsst.table.io`` framework.
+However, ``lsst.afw.typehandling`` has no other dependencies on ``afw``.
+Once object persistence is decoupled from ``lsst.table.io``, the entire subpackage can be moved to a lower level in the Stack and :class:`~lsst.afw.typehandling.GenericMap` can be treated as a fundamental LSST type.
+
+Design Goals
+------------
+
+We designed and implemented :class:`~lsst.afw.typehandling.GenericMap` while striving to obey the following principles:
+
+* provide both a C++ and a Python API for :class:`~lsst.afw.typehandling.GenericMap`, although it is unlikely that many Python users will use :class:`~lsst.afw.typehandling.GenericMap` directly, rather through classes such as :class:`~lsst.afw.image.ExposureInfo`.
+* support storage of primitive types, as well as LSST classes written in either C++ or Python.
+* provide C++ and Python APIs that are as idiomatic as possible in their respective languages, making full use of pybind11's ability to serve as an API adapter.
+* do not provide any features beyond a heterogeneous mapping with simple set-get behavior.
+  The single responsibility makes :class:`~lsst.afw.typehandling.GenericMap` suitable as a back-end for a variety of applications, including more ornate mapping types like :class:`~lsst.daf.base.PropertySet`.
+* provide separate interface (:class:`~lsst.afw.typehandling.GenericMap` and :class:`~lsst.afw.typehandling.MutableGenericMap`) and implementation classes.
+  An abstract interface makes it easy to create mappings that require specific properties (as :class:`~lsst.daf.base.PropertySet` and :class:`~lsst.daf.base.PropertyList` do) without breaking code for other users, following the open-closed principle.
+  Most code based on :class:`~lsst.afw.typehandling.GenericMap` can be agnostic to the implementation class.
+* rely on compile-time type safety as much as possible, with safeguards preventing invalid element retrieval in C++ and spurious type errors in Python.
+* allow element retrieval without knowing the exact type as which it was stored, so long as the conversion is valid (e.g., superclass vs. subclass, or different sizes of floating-point number).
+  Supporting inexact types is not only user-friendly, it avoids unnecessary coupling between code changes at the points of storage and retrieval (which may be in different packages).
```

Comment:
Being _especially_ picky, it bothers me that these bullet points start with lower-case letters, while those in §2 above start with upper-case letters.